### INTRO - 
<b> Refer One Note </b>

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [3]:
autos = pd.read_csv('autos.csv', encoding='Latin-1')

In [4]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [5]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

In [15]:
autos.describe(include='all').transpose()

count unique                  top   freq     mean  \
dateCrawled          50000  48213  2016-04-04 16:40:33      3      NaN   
name                 50000  38754          Ford_Fiesta     78      NaN   
seller               50000      2               privat  49999      NaN   
offerType            50000      2              Angebot  49999      NaN   
price                50000   2357                   $0   1421      NaN   
abtest               50000      2                 test  25756      NaN   
vehicleType          44905      8            limousine  12859      NaN   
yearOfRegistration   50000    NaN                  NaN    NaN  2005.07   
gearbox              47320      2              manuell  36993      NaN   
powerPS              50000    NaN                  NaN    NaN  116.356   
model                47242    245                 golf   4024      NaN   
odometer             50000     13            150,000km  32424      NaN   
monthOfRegistration  50000    NaN                  NaN    NaN  5.72336   
fuelType             45518      7               benzin  30107      NaN   
brand                50000     40           volkswagen  10687      NaN   
notRepairedDamage    40171      2                 nein  35232      NaN   
dateCreated          50000     76  2016-04-03 00:00:00   1946      NaN   
nrOfPictures         50000    NaN                  NaN    NaN        0   
postalCode           50000    NaN                  NaN    NaN  50813.6   
lastSeen             50000  39481  2016-04-07 06:17:27      8      NaN   

                         std   min    25%    50%    75%    max  
dateCrawled              NaN   NaN    NaN    NaN    NaN    NaN  
name                     NaN   NaN    NaN    NaN    NaN    NaN  
seller                   NaN   NaN    NaN    NaN    NaN    NaN  
offerType                NaN   NaN    NaN    NaN    NaN    NaN  
price                    NaN   NaN    NaN    NaN    NaN    NaN  
abtest                   NaN   NaN    NaN    NaN    NaN    NaN  
vehicleType              NaN   NaN    NaN    NaN    NaN    NaN  
yearOfRegistration   105.713  1000   1999   2003   2008   9999  
gearbox                  NaN   NaN    NaN    NaN    NaN    NaN  
powerPS              209.217     0     70    105    150  17700  
model                    NaN   NaN    NaN    NaN    NaN    NaN  
odometer                 NaN   NaN    NaN    NaN    NaN    NaN  
monthOfRegistration  3.71198     0      3      6      9     12  
fuelType                 NaN   NaN    NaN    NaN    NaN    NaN  
brand                    NaN   NaN    NaN    NaN    NaN    NaN  
notRepairedDamage        NaN   NaN    NaN    NaN    NaN    NaN  
dateCreated              NaN   NaN    NaN    NaN    NaN    NaN  
nrOfPictures               0     0      0      0      0      0  
postalCode           25779.7  1067  30451  49577  71540  99998  
lastSeen                 NaN   NaN    NaN    NaN    NaN    NaN

### Clean Up Column names

- We convert column names from Camel to snake case
- also rename few column to more appropriate names

In [16]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [17]:
autos.rename(columns = {'dateCrawled':'date_crawled', 'offerType':'offer_type',
                        'powerPS':'power_ps','vehicleType':'vehicle_type',
                        'dateCreated':'ad_created',
                        'fuelType':'fuel_type',
                        'nrOfPictures':'no_of_pictures',
                        'postalCode':'postal_code',
                        'lastSeen':'last_seen',
                        'notRepairedDamage' : 'unrepaired_damage',
                        'yearOfRegistration' : 'registration_year', 
                        'monthOfRegistration' : 'registration_month', 
                       },inplace=True)

In [18]:
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'no_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

In [19]:
autos.describe(include='all').transpose()

count unique                  top   freq     mean  \
date_crawled        50000  48213  2016-04-04 16:40:33      3      NaN   
name                50000  38754          Ford_Fiesta     78      NaN   
seller              50000      2               privat  49999      NaN   
offer_type          50000      2              Angebot  49999      NaN   
price               50000   2357                   $0   1421      NaN   
abtest              50000      2                 test  25756      NaN   
vehicle_type        44905      8            limousine  12859      NaN   
registration_year   50000    NaN                  NaN    NaN  2005.07   
gearbox             47320      2              manuell  36993      NaN   
power_ps            50000    NaN                  NaN    NaN  116.356   
model               47242    245                 golf   4024      NaN   
odometer            50000     13            150,000km  32424      NaN   
registration_month  50000    NaN                  NaN    NaN  5.72336   
fuel_type           45518      7               benzin  30107      NaN   
brand               50000     40           volkswagen  10687      NaN   
unrepaired_damage   40171      2                 nein  35232      NaN   
ad_created          50000     76  2016-04-03 00:00:00   1946      NaN   
no_of_pictures      50000    NaN                  NaN    NaN        0   
postal_code         50000    NaN                  NaN    NaN  50813.6   
last_seen           50000  39481  2016-04-07 06:17:27      8      NaN   

                        std   min    25%    50%    75%    max  
date_crawled            NaN   NaN    NaN    NaN    NaN    NaN  
name                    NaN   NaN    NaN    NaN    NaN    NaN  
seller                  NaN   NaN    NaN    NaN    NaN    NaN  
offer_type              NaN   NaN    NaN    NaN    NaN    NaN  
price                   NaN   NaN    NaN    NaN    NaN    NaN  
abtest                  NaN   NaN    NaN    NaN    NaN    NaN  
vehicle_type            NaN   NaN    NaN    NaN    NaN    NaN  
registration_year   105.713  1000   1999   2003   2008   9999  
gearbox                 NaN   NaN    NaN    NaN    NaN    NaN  
power_ps            209.217     0     70    105    150  17700  
model                   NaN   NaN    NaN    NaN    NaN    NaN  
odometer                NaN   NaN    NaN    NaN    NaN    NaN  
registration_month  3.71198     0      3      6      9     12  
fuel_type               NaN   NaN    NaN    NaN    NaN    NaN  
brand                   NaN   NaN    NaN    NaN    NaN    NaN  
unrepaired_damage       NaN   NaN    NaN    NaN    NaN    NaN  
ad_created              NaN   NaN    NaN    NaN    NaN    NaN  
no_of_pictures            0     0      0      0      0      0  
postal_code         25779.7  1067  30451  49577  71540  99998  
last_seen               NaN   NaN    NaN    NaN    NaN    NaN

In [20]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   date_crawled        50000 non-null  object
 1   name                50000 non-null  object
 2   seller              50000 non-null  object
 3   offer_type          50000 non-null  object
 4   price               50000 non-null  object
 5   abtest              50000 non-null  object
 6   vehicle_type        44905 non-null  object
 7   registration_year   50000 non-null  int64 
 8   gearbox             47320 non-null  object
 9   power_ps            50000 non-null  int64 
 10  model               47242 non-null  object
 11  odometer            50000 non-null  object
 12  registration_month  50000 non-null  int64 
 13  fuel_type           45518 non-null  object
 14  brand               50000 non-null  object
 15  unrepaired_damage   40171 non-null  object
 16  ad_created          50

In [21]:
autos.price.head(2)       # Need to convert these to Numeric

0    $5,000
1    $8,500
Name: price, dtype: object

In [22]:
autos.odometer.head(2)         # Need to convert these to Numeric

0    150,000km
1    150,000km
Name: odometer, dtype: object

In [23]:
autos.seller.value_counts()                 # Can drop this column

privat        49999
gewerblich        1
Name: seller, dtype: int64

In [24]:
autos.offer_type.value_counts()            # Can drop this column

Angebot    49999
Gesuch         1
Name: offer_type, dtype: int64

In [25]:
autos.no_of_pictures.value_counts()          #

0    50000
Name: no_of_pictures, dtype: int64

In [26]:
autos.postal_code.value_counts()

10115    109
65428    104
66333     54
45888     50
44145     48
        ... 
23942      1
83365      1
95683      1
97794      1
67585      1
Name: postal_code, Length: 7014, dtype: int64

### Data Investigation / Cleaning  -- Numeric Columns

- Convert `price` to Numeric data type
- Convert `odometer` to Numeric data type
- We'll also check for outliers (unusually high or low values), nulls etc. on these columns 



- Also, We can potentially drop `seller`,`offer_type`, `no_of_Pictures`. from our analysis since only a single value exists for all rows

In [27]:
autos.price.head(2)

0    $5,000
1    $8,500
Name: price, dtype: object

In [28]:
autos.price = (autos.price.str.strip('$').str.split(',').str[0])+(autos.price.str.strip('$').str.split(',').str[1])

In [29]:
autos.price = autos.price.astype('float')

In [30]:
autos.price.head(2)

0    5000.0
1    8500.0
Name: price, dtype: float64

In [31]:
autos.odometer.head(2)

0    150,000km
1    150,000km
Name: odometer, dtype: object

In [32]:
autos.odometer = (autos.odometer.str.strip('km').str.split(',').str[0])+(autos.odometer.str.strip('km').str.split(',').str[1])

In [33]:
autos.odometer.head(2)

0    150000
1    150000
Name: odometer, dtype: object

In [34]:
autos.odometer = autos.odometer.astype('float')

In [35]:
autos.odometer.head(2)

0    150000.0
1    150000.0
Name: odometer, dtype: float64

In [36]:
# Identifying any outliers on Price Column
# Unusually high or low values

autos.price.value_counts().unique().shape      #169 unique prices

(169,)

In [37]:
autos.price.describe()

# Price of 999,999 seems like a mistake
# Also, we see $1000 car ads which is possible so we can ignore that

count     38640.000000
mean       7335.640631
std       13068.127589
min        1000.000000
25%        2200.000000
50%        4350.000000
75%        8950.000000
max      999999.000000
Name: price, dtype: float64

In [38]:
autos.price.value_counts().sort_index(ascending=True).tail(10)

# Only last 2 entries here of 999,999 seem like errors

220000.0    1
250000.0    1
259000.0    1
265000.0    1
295000.0    1
299000.0    1
345000.0    1
350000.0    1
999990.0    1
999999.0    2
Name: price, dtype: int64

In [39]:
autos.loc[autos.price>=999990.0,:]       #Problematic columns

date_crawled                              name  seller  \
514    2016-03-17 09:53:08  Ford_Focus_Turnier_1.6_16V_Style  privat   
37585  2016-03-29 11:38:54               Volkswagen_Jetta_GT  privat   
43049  2016-03-21 19:53:52                     2_VW_Busse_T3  privat   

      offer_type     price abtest vehicle_type  registration_year  gearbox  \
514      Angebot  999999.0   test        kombi               2009  manuell   
37585    Angebot  999990.0   test    limousine               1985  manuell   
43049    Angebot  999999.0   test          bus               1981  manuell   

       power_ps        model  odometer  registration_month fuel_type  \
514         101        focus  125000.0                   4    benzin   
37585       111        jetta  150000.0                  12    benzin   
43049        70  transporter  150000.0                   1    benzin   

            brand unrepaired_damage           ad_created  no_of_pictures  \
514          ford              nein  2016-03-17 00:00:00               0   
37585  volkswagen                ja  2016-03-29 00:00:00               0   
43049  volkswagen               NaN  2016-03-21 00:00:00               0   

       postal_code            last_seen  
514          12205  2016-04-06 07:17:35  
37585        50997  2016-03-29 11:38:54  
43049        99880  2016-03-28 17:18:28

In [40]:
autos.odometer.value_counts().unique().shape     #13 unique Odometer readings

(13,)

In [41]:
autos.odometer.describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer, dtype: float64

In [42]:
autos.odometer.value_counts().sort_index()     #For now nothing seems like a outlier as such

5000.0        967
10000.0       264
20000.0       784
30000.0       789
40000.0       819
50000.0      1027
60000.0      1164
70000.0      1230
80000.0      1436
90000.0      1757
100000.0     2169
125000.0     5170
150000.0    32424
Name: odometer, dtype: int64

In [43]:
### DROP ANY UNWANTED COLUMNS
### Drop Any Outlier ROWS

#### [PENDING] Following clean up above, make some observations on odometer and price

### Data Investigation / Cleaning -- Date columns

Date Columns:
- `date_crawled`: added by the crawler
- `last_seen`: added by the crawler
- `ad_created`: from the website
- `registration_month`: from the website
- `registration_year`: from the website


In [44]:
autos.loc[:, ['date_crawled','last_seen','ad_created','registration_month','registration_year']].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 5 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   date_crawled        50000 non-null  object
 1   last_seen           50000 non-null  object
 2   ad_created          50000 non-null  object
 3   registration_month  50000 non-null  int64 
 4   registration_year   50000 non-null  int64 
dtypes: int64(2), object(3)
memory usage: 1.9+ MB


In [45]:
# date crawled, last seen and ad_created are of type object and need to be converted in order to analyze

In [46]:
autos[['date_crawled','ad_created','last_seen']].head(5)

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

In [47]:
autos.date_crawled.str[:10].value_counts(normalize=True,dropna = False).sort_index()


2016-03-05    0.02538
2016-03-06    0.01394
2016-03-07    0.03596
2016-03-08    0.03330
2016-03-09    0.03322
2016-03-10    0.03212
2016-03-11    0.03248
2016-03-12    0.03678
2016-03-13    0.01556
2016-03-14    0.03662
2016-03-15    0.03398
2016-03-16    0.02950
2016-03-17    0.03152
2016-03-18    0.01306
2016-03-19    0.03490
2016-03-20    0.03782
2016-03-21    0.03752
2016-03-22    0.03294
2016-03-23    0.03238
2016-03-24    0.02910
2016-03-25    0.03174
2016-03-26    0.03248
2016-03-27    0.03104
2016-03-28    0.03484
2016-03-29    0.03418
2016-03-30    0.03362
2016-03-31    0.03192
2016-04-01    0.03380
2016-04-02    0.03540
2016-04-03    0.03868
2016-04-04    0.03652
2016-04-05    0.01310
2016-04-06    0.00318
2016-04-07    0.00142
Name: date_crawled, dtype: float64

In [48]:
autos.ad_created.str[:10].value_counts(normalize=True,dropna = False).sort_index()


2015-06-11    0.00002
2015-08-10    0.00002
2015-09-09    0.00002
2015-11-10    0.00002
2015-12-05    0.00002
               ...   
2016-04-03    0.03892
2016-04-04    0.03688
2016-04-05    0.01184
2016-04-06    0.00326
2016-04-07    0.00128
Name: ad_created, Length: 76, dtype: float64

In [49]:
autos.last_seen.str[:10].value_counts(normalize=True,dropna = False).sort_index()


2016-03-05    0.00108
2016-03-06    0.00442
2016-03-07    0.00536
2016-03-08    0.00760
2016-03-09    0.00986
2016-03-10    0.01076
2016-03-11    0.01252
2016-03-12    0.02382
2016-03-13    0.00898
2016-03-14    0.01280
2016-03-15    0.01588
2016-03-16    0.01644
2016-03-17    0.02792
2016-03-18    0.00742
2016-03-19    0.01574
2016-03-20    0.02070
2016-03-21    0.02074
2016-03-22    0.02158
2016-03-23    0.01858
2016-03-24    0.01956
2016-03-25    0.01920
2016-03-26    0.01696
2016-03-27    0.01602
2016-03-28    0.02086
2016-03-29    0.02234
2016-03-30    0.02484
2016-03-31    0.02384
2016-04-01    0.02310
2016-04-02    0.02490
2016-04-03    0.02536
2016-04-04    0.02462
2016-04-05    0.12428
2016-04-06    0.22100
2016-04-07    0.13092
Name: last_seen, dtype: float64

In [50]:
autos.registration_year.describe()

count    50000.000000
mean      2005.073280
std        105.712813
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

### [Pending] Observations on the Date columns


- We see that registration year contains some odd values (like max is 9999 and min year is 1000).
need to investigate it further


- takeaway 2
 
 ...
 

In [51]:
#Checking registration year

autos.loc[(autos.registration_year<1900) | (autos.registration_year>2016),'registration_year'].value_counts()

2017    1453
2018     492
9999       4
5000       4
2019       3
9000       2
1800       2
6200       1
4500       1
8888       1
4800       1
2800       1
1001       1
1000       1
1111       1
1500       1
9996       1
5911       1
4100       1
Name: registration_year, dtype: int64

<b> Cleaning up registration year </b>
The latest year on the last_seen column was 2016, Its not possible to have a car registered after its Ad was last seen. So we can drop rows where registration year was greater and 2016. Also we'll drop rows for cars earlier than year 1900.

In [52]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   date_crawled        50000 non-null  object 
 1   name                50000 non-null  object 
 2   seller              50000 non-null  object 
 3   offer_type          50000 non-null  object 
 4   price               38640 non-null  float64
 5   abtest              50000 non-null  object 
 6   vehicle_type        44905 non-null  object 
 7   registration_year   50000 non-null  int64  
 8   gearbox             47320 non-null  object 
 9   power_ps            50000 non-null  int64  
 10  model               47242 non-null  object 
 11  odometer            50000 non-null  float64
 12  registration_month  50000 non-null  int64  
 13  fuel_type           45518 non-null  object 
 14  brand               50000 non-null  object 
 15  unrepaired_damage   40171 non-null  object 
 16  ad_c

In [53]:
#To get Index of rows to delete

autos.loc[(autos.registration_year<1900) | (autos.registration_year>2016)].index

Int64Index([   10,    55,    65,    68,    84,   113,   164,   197,   253,
              348,
            ...
            49662, 49689, 49696, 49731, 49770, 49796, 49841, 49880, 49910,
            49935],
           dtype='int64', length=1972)

In [54]:
#drop the rows

autos.drop(labels= \
           autos.loc[(autos.registration_year<1900) | (autos.registration_year>2016)].index, inplace=True)

In [55]:
autos.info()
#We see that number of rows is now reduced

<class 'pandas.core.frame.DataFrame'>
Int64Index: 48028 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   date_crawled        48028 non-null  object 
 1   name                48028 non-null  object 
 2   seller              48028 non-null  object 
 3   offer_type          48028 non-null  object 
 4   price               37218 non-null  float64
 5   abtest              48028 non-null  object 
 6   vehicle_type        44903 non-null  object 
 7   registration_year   48028 non-null  int64  
 8   gearbox             45604 non-null  object 
 9   power_ps            48028 non-null  int64  
 10  model               45560 non-null  object 
 11  odometer            48028 non-null  float64
 12  registration_month  48028 non-null  int64  
 13  fuel_type           44301 non-null  object 
 14  brand               48028 non-null  object 
 15  unrepaired_damage   39040 non-null  object 
 16  ad_c

In [56]:
autos.registration_year.value_counts().sort_index()

1910       9
1927       1
1929       1
1931       1
1934       2
        ... 
2012    1323
2013     806
2014     666
2015     399
2016    1316
Name: registration_year, Length: 78, dtype: int64

<b> Analyzing the `brand` columns </b>
- we will look at the average price of top 15 car brands listed on the website

In [57]:
autos.brand.head(2)

0    peugeot
1        bmw
Name: brand, dtype: object

In [58]:
top_brands_by_count = autos.brand.value_counts().sort_values(ascending=False).head(15).index

In [59]:
top_brands_by_count

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat', 'skoda', 'mazda', 'nissan', 'citroen',
       'smart'],
      dtype='object')

In [60]:
autos.price

0         5000.0
1         8500.0
2         8990.0
3         4350.0
4         1350.0
          ...   
49995    24900.0
49996     1980.0
49997    13200.0
49998    22900.0
49999     1250.0
Name: price, Length: 48028, dtype: float64

In [61]:
autos.brand

0           peugeot
1               bmw
2        volkswagen
3             smart
4              ford
            ...    
49995          audi
49996          opel
49997          fiat
49998          audi
49999          opel
Name: brand, Length: 48028, dtype: object

In [62]:
autos.loc[autos.brand=='bmw','price'].mean().round(2)

9117.51

In [71]:
brand_mean_prices = {}

for car_brand in top_brands_by_count:
    brand_mean_prices[car_brand] = autos.loc[autos.brand==car_brand,'price'].mean().round(2)

print(brand_mean_prices)


{'volkswagen': 6898.91, 'bmw': 9117.51, 'opel': 4222.41, 'mercedes_benz': 9324.45, 'audi': 10322.27, 'ford': 5789.14, 'renault': 3590.94, 'peugeot': 3955.17, 'fiat': 4008.17, 'seat': 5638.64, 'skoda': 6836.7, 'mazda': 5309.53, 'nissan': 6428.43, 'citroen': 4658.81, 'smart': 3780.47}


#### [PENDING]Takeaways on Price

 takeaway 1
 
 takeaway 2
 
 ...
 





### Looking at any relation between  avg. mileage and avg .price

In [76]:
autos.brand.unique()

array(['peugeot', 'bmw', 'volkswagen', 'smart', 'ford', 'chrysler',
       'seat', 'renault', 'mercedes_benz', 'audi', 'sonstige_autos',
       'opel', 'mazda', 'porsche', 'mini', 'toyota', 'dacia', 'nissan',
       'jeep', 'saab', 'volvo', 'mitsubishi', 'jaguar', 'fiat', 'skoda',
       'subaru', 'kia', 'citroen', 'chevrolet', 'hyundai', 'honda',
       'daewoo', 'suzuki', 'trabant', 'land_rover', 'alfa_romeo', 'lada',
       'rover', 'daihatsu', 'lancia'], dtype=object)

In [92]:
#autos.loc[:,'mileage']
autos.odometer

brand_mean_odo = {}

for brand_name in autos.brand.unique():
    brand_mean_odo[brand_name] = autos.loc[autos.brand==brand_name,"odometer"].mean().round(2)

print(brand_mean_odo)

{'peugeot': 127136.81, 'bmw': 132434.71, 'volkswagen': 128730.37, 'smart': 99595.81, 'ford': 124046.84, 'chrysler': 133181.82, 'seat': 121563.57, 'renault': 128183.82, 'mercedes_benz': 130860.26, 'audi': 129287.78, 'sonstige_autos': 87262.36, 'opel': 129227.14, 'mazda': 124745.53, 'porsche': 97457.34, 'mini': 88602.41, 'toyota': 115709.52, 'dacia': 84268.29, 'nissan': 118572.41, 'jeep': 127546.3, 'saab': 144415.58, 'volvo': 138355.86, 'mitsubishi': 126930.95, 'jaguar': 120921.05, 'fiat': 116553.95, 'skoda': 110954.55, 'subaru': 124857.14, 'kia': 112434.02, 'citroen': 119461.88, 'chevrolet': 99251.82, 'hyundai': 106511.63, 'honda': 122851.46, 'daewoo': 122430.56, 'suzuki': 109049.3, 'trabant': 59666.67, 'land_rover': 118010.2, 'alfa_romeo': 131399.37, 'lada': 85517.24, 'rover': 135615.38, 'daihatsu': 115284.55, 'lancia': 122019.23}


In [94]:
#Convert both dictionaries to Pandas series

bmp_series = pd.Series(brand_mean_prices)

odo_series = pd.Series(brand_mean_odo)

In [95]:
bmp_series

volkswagen        6898.91
bmw               9117.51
opel              4222.41
mercedes_benz     9324.45
audi             10322.27
ford              5789.14
renault           3590.94
peugeot           3955.17
fiat              4008.17
seat              5638.64
skoda             6836.70
mazda             5309.53
nissan            6428.43
citroen           4658.81
smart             3780.47
dtype: float64

In [96]:
brand_df = pd.DataFrame(bmp_series,columns=['mean_price'])

In [97]:
brand_df.head()

mean_price
volkswagen        6898.91
bmw               9117.51
opel              4222.41
mercedes_benz     9324.45
audi             10322.27

In [98]:
odo_series

peugeot           127136.81
bmw               132434.71
volkswagen        128730.37
smart              99595.81
ford              124046.84
chrysler          133181.82
seat              121563.57
renault           128183.82
mercedes_benz     130860.26
audi              129287.78
sonstige_autos     87262.36
opel              129227.14
mazda             124745.53
porsche            97457.34
mini               88602.41
toyota            115709.52
dacia              84268.29
nissan            118572.41
jeep              127546.30
saab              144415.58
volvo             138355.86
mitsubishi        126930.95
jaguar            120921.05
fiat              116553.95
skoda             110954.55
subaru            124857.14
kia               112434.02
citroen           119461.88
chevrolet          99251.82
hyundai           106511.63
honda             122851.46
daewoo            122430.56
suzuki            109049.30
trabant            59666.67
land_rover        118010.20
alfa_romeo        13

In [99]:
brand_df['mean_mileage'] = odo_series

In [100]:
brand_df.head()

mean_price  mean_mileage
volkswagen        6898.91     128730.37
bmw               9117.51     132434.71
opel              4222.41     129227.14
mercedes_benz     9324.45     130860.26
audi             10322.27     129287.78

#### [PENDING]Takeaways on Price vs Mileage

 takeaway 1
 
 takeaway 2
 
 ...
 

<b>[PENDING] Any Final Conclusions on the analysis </b>

refer for solution/takeaways:

https://github.com/dataquestio/solutions/blob/master/Mission294Solutions.ipynb